File: src/lib/crypto.ts
// Uses Node.js built-in crypto — no extra dependency needed
import crypto from "crypto";

const ALGO = "aes-256-gcm";

function getEncryptionKey(): Buffer {
  const rawKey = process.env.ENCRYPTION_KEY || "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
  // SHA-256 hash guarantees an exact 32-byte (256-bit) buffer regardless of rawKey length or format
  return crypto.createHash("sha256").update(rawKey).digest();
}

export function encrypt(plaintext: string): string {
  const key = getEncryptionKey();
  const iv = crypto.randomBytes(16);
  const cipher = crypto.createCipheriv(ALGO, key, iv);
  const encrypted = Buffer.concat([
    cipher.update(plaintext, "utf8"),
    cipher.final(),
  ]);
  const tag = cipher.getAuthTag();
  // Store iv:tag:ciphertext as base64
  return [iv, tag, encrypted].map((b) => b.toString("base64")).join(":");
}

export function decrypt(stored: string): string {
  const [ivB64, tagB64, encB64] = stored.split(":");
  if (!ivB64 || !tagB64 || !encB64) {
    throw new Error("DECRYPTION_FAILED: Invalid stored format");
  }
  const iv        = Buffer.from(ivB64,  "base64");
  const tag       = Buffer.from(tagB64, "base64");
  const encrypted = Buffer.from(encB64, "base64");

  try {
    const key = getEncryptionKey();
    const decipher  = crypto.createDecipheriv(ALGO, key, iv);
    decipher.setAuthTag(tag);
    return decipher.update(encrypted) + decipher.final("utf8");
  } catch (err) {
    try {
      // Fallback to default dev key if primary decryption fails
      const fallbackRaw = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef";
      const fallbackKey = crypto.createHash("sha256").update(fallbackRaw).digest();
      const decipher  = crypto.createDecipheriv(ALGO, fallbackKey, iv);
      decipher.setAuthTag(tag);
      return decipher.update(encrypted) + decipher.final("utf8");
    } catch (fallbackErr) {
      throw new Error("DECRYPTION_FAILED: Decryption key mismatch");
    }
  }
}

File: services/transaction_syncer/engine/crypto.py
import base64
import hashlib
import os
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

DEFAULT_KEY_RAW = "0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"

def get_encryption_key(raw_key: str = None) -> bytes:
    key_str = raw_key or os.getenv("ENCRYPTION_KEY", DEFAULT_KEY_RAW)
    return hashlib.sha256(key_str.encode("utf-8")).digest()

def decrypt(stored: str, raw_key: str = None) -> str:
    if not stored:
        return ""
    parts = stored.split(":")
    if len(parts) != 3:
        # Unencrypted plaintext or fallback
        return stored
    
    iv_b64, tag_b64, enc_b64 = parts
    try:
        iv = base64.b64decode(iv_b64)
        tag = base64.b64decode(tag_b64)
        ciphertext = base64.b64decode(enc_b64)

        # In Python cryptography AESGCM, payload is ciphertext + tag
        data_to_decrypt = ciphertext + tag
        
        key = get_encryption_key(raw_key)
        try:
            aesgcm = AESGCM(key)
            return aesgcm.decrypt(iv, data_to_decrypt, None).decode("utf-8")
        except Exception:
            try:
                fallback_key = get_encryption_key(DEFAULT_KEY_RAW)
                aesgcm_fb = AESGCM(fallback_key)
                return aesgcm_fb.decrypt(iv, data_to_decrypt, None).decode("utf-8")
            except Exception:
                return stored
    except Exception:
        return stored

def encrypt(plaintext: str, raw_key: str = None) -> str:
    key = get_encryption_key(raw_key)
    aesgcm = AESGCM(key)
    iv = os.urandom(16)
    encrypted_payload = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
    ciphertext = encrypted_payload[:-16]
    tag = encrypted_payload[-16:]
    
    iv_b64 = base64.b64encode(iv).decode("utf-8")
    tag_b64 = base64.b64encode(tag).decode("utf-8")
    enc_b64 = base64.b64encode(ciphertext).decode("utf-8")
    return f"{iv_b64}:{tag_b64}:{enc_b64}"
